### DISCUSSION TITLE
RTL SDR USB Buffers?

### DISCUSSION THREAD

**Discussion by WaterstarRunner** [jvde-github/AIS-catcher#186] [General]
Hiya,

I notice that the command line BUFFER_COUNT and JSON conf buffer_count for RTLSDR don't work / aren't documented, but do work if RTLSDR.cpp is modified from 0 (default 15 buffers) to BUFFER_COUNT (24 from RTLSDR.h) or as user configured...

`rtlsdr_read_async(dev, (rtlsdr_read_async_cb_t) & (RTLSDR::callback_static), this, BUFFER_COUNT, BUFFER_SIZE);`

Is there any reason why the configurability on this is disabled?

Not a big problem, I was trying to solve for multiple sdr receivers in multiple daemons crashing, but it turned out to be because my USB power consumption was maxed out, not a shortage of USB buffers.

Just curious. I have no idea what I'm doing, but my build with 12 default in RTLSDR.h and 8 in conf works nicely...

**Comment by jvde-github** [comment]
Hi, the BUFFER_COUNT is used [here](https://github.com/jvde-github/AIS-catcher/blob/1e2400501cdb7f64d6ec3b638bff552497bf330f/Device/RTLSDR.cpp#L56). For the additional buffering I am not using the functionality from the native RTL-SDR driver but a built-in buffer (called FIFO) that works on top of this.  So you will see in the output that RTL-SDR reports it is using the default buffers but not the additional buffers that are used (only when it reports out the value of BUFFER_COUNT). So in short, there are two buffers :-) Does that make sense?

Cheers

**Reply by WaterstarRunner** [reply]
Ah, fantastic, thanks so much for sharing (and for building this in the first place :)) 

At some point I'll have to make an attempt to understand the buffer types, purposes, and resource contention... but in the meantime, my last bottleneck for number of active RTLSDRs on host was the power supply, and the current bottleneck is a shortage of antennas...

**Reply by jvde-github** [reply]
Great! If you run into performance issue, would first try to use the -F switch to speed up downsampling. If that works but you have occasional buffer overruns, I would increase the BUFFER_COUNT.

Cheers
